File: backend/app/schemas/work_item.py
from __future__ import annotations

from datetime import date, datetime
from typing import Literal
from zoneinfo import ZoneInfo, ZoneInfoNotFoundError

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator

WorkItemType = Literal["task", "order", "realization", "reminder", "event"]
WorkItemStatus = Literal["todo", "in_progress", "completed", "cancelled"]
WorkItemPriority = Literal["low", "normal", "high", "urgent"]
CandidateResolution = Literal["new_candidate"]
AbsenceStatus = Literal["requested", "approved", "rejected", "cancelled"]


class WorkItemCreate(BaseModel):
    item_type: WorkItemType
    title: str = Field(min_length=1, max_length=255)
    description: str | None = Field(default=None, max_length=20000)
    start_at: datetime | None = None
    due_at: datetime | None = None
    all_day: bool = False
    timezone_name: str | None = Field(default=None, max_length=64)
    status: WorkItemStatus = "todo"
    priority: WorkItemPriority = "normal"
    assignee_user_id: int | None = Field(default=None, gt=0)
    client_id: int | None = Field(default=None, gt=0)
    party_name: str | None = Field(default=None, max_length=255)

    @field_validator("title", "description", "timezone_name", "party_name", mode="before")
    @classmethod
    def strip_text(cls, value):
        if isinstance(value, str):
            value = value.strip()
            return value or None
        return value

    @field_validator("start_at", "due_at")
    @classmethod
    def require_timezone(cls, value):
        if value is not None and value.tzinfo is None:
            raise ValueError("date-time values must include a timezone offset")
        return value

    @model_validator(mode="after")
    def validate_semantics(self):
        if self.due_at and self.start_at and self.due_at < self.start_at:
            raise ValueError("due_at cannot precede start_at")
        if self.item_type == "event" and self.start_at is None:
            raise ValueError("event requires start_at")
        if self.item_type == "reminder" and self.due_at is None:
            raise ValueError("reminder requires due_at")
        if self.all_day and (self.start_at is None or self.due_at is None or not self.timezone_name):
            raise ValueError("all_day requires start_at, due_at and timezone_name")
        if self.timezone_name:
            try:
                ZoneInfo(self.timezone_name)
            except ZoneInfoNotFoundError as error:
                raise ValueError("timezone_name must be a valid IANA timezone") from error
        return self


class WorkItemUpdate(BaseModel):
    expected_version: int = Field(gt=0)
    item_type: WorkItemType | None = None
    title: str | None = Field(default=None, min_length=1, max_length=255)
    description: str | None = Field(default=None, max_length=20000)
    start_at: datetime | None = None
    due_at: datetime | None = None
    all_day: bool | None = None
    timezone_name: str | None = Field(default=None, max_length=64)
    status: WorkItemStatus | None = None
    priority: WorkItemPriority | None = None
    assignee_user_id: int | None = Field(default=None, gt=0)
    client_id: int | None = Field(default=None, gt=0)
    party_name: str | None = Field(default=None, max_length=255)

    @field_validator("start_at", "due_at")
    @classmethod
    def require_timezone(cls, value):
        if value is not None and value.tzinfo is None:
            raise ValueError("date-time values must include a timezone offset")
        return value


class WorkItemStatusUpdate(BaseModel):
    status: WorkItemStatus
    expected_version: int = Field(gt=0)


class VersionRequest(BaseModel):
    expected_version: int = Field(gt=0)


class WorkItemRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    item_type: WorkItemType
    title: str
    description: str | None
    start_at: datetime | None
    due_at: datetime | None
    all_day: bool
    timezone_name: str | None
    status: WorkItemStatus
    priority: WorkItemPriority
    assignee_user_id: int | None
    assignee_display: str | None = None
    client_id: int | None
    client_name: str | None = None
    party_name: str | None
    created_by_user_id: int
    updated_by_user_id: int
    created_at: datetime
    updated_at: datetime
    completed_at: datetime | None
    deleted_at: datetime | None
    version: int


class WorkItemPage(BaseModel):
    items: list[WorkItemRead]
    total: int
    skip: int
    limit: int


class WorkItemNoteCreate(BaseModel):
    text: str = Field(min_length=1, max_length=10000)

    @field_validator("text")
    @classmethod
    def strip_text(cls, value: str) -> str:
        value = value.strip()
        if not value:
            raise ValueError("note text is required")
        return value


class WorkItemNoteUpdate(WorkItemNoteCreate):
    expected_version: int = Field(gt=0)


class WorkItemNoteRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    work_item_id: int
    text: str
    created_by_user_id: int
    updated_by_user_id: int
    created_at: datetime
    updated_at: datetime
    deleted_at: datetime | None
    version: int


class WorkItemDocumentLink(BaseModel):
    document_id: int = Field(gt=0)
    note_id: int | None = Field(default=None, gt=0)


class WorkItemDocumentRead(BaseModel):
    id: int
    work_item_id: int
    note_id: int | None
    document_id: int
    filename: str
    content_type: str
    file_size: int
    source_type: str
    captured_at: datetime | None
    created_at: datetime


class AbsenceCreate(BaseModel):
    absence_type: Literal["vacation", "day_off", "sick_leave", "other"]
    start_date: date
    end_date: date
    note: str | None = Field(default=None, max_length=5000)

    @model_validator(mode="after")
    def validate_range(self):
        if self.end_date < self.start_date:
            raise ValueError("end_date cannot precede start_date")
        return self


class AbsenceUpdate(AbsenceCreate):
    expected_version: int = Field(gt=0)


class AbsenceReview(BaseModel):
    expected_version: int = Field(gt=0)
    review_note: str | None = Field(default=None, max_length=2000)


class AbsenceRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    requester_user_id: int
    requester_display: str | None = None
    absence_type: Literal["vacation", "day_off", "sick_leave", "other"]
    start_date: date
    end_date: date
    status: AbsenceStatus
    note: str | None
    reviewed_by_user_id: int | None
    reviewed_at: datetime | None
    review_note: str | None
    cancelled_by_user_id: int | None
    cancelled_at: datetime | None
    created_at: datetime
    updated_at: datetime
    version: int


class AbsencePage(BaseModel):
    items: list[AbsenceRead]
    total: int
    skip: int
    limit: int


class CalendarEntry(BaseModel):
    entity_id: int
    entity_kind: Literal["work_item", "absence"]
    item_type: str
    title: str
    start: datetime | date
    end: datetime | date
    status: str
    priority: str | None = None
    assignee_display: str | None = None
    client_id: int | None = None
    client_name: str | None = None
    all_day: bool


class CalendarMonth(BaseModel):
    year: int
    month: int
    items: list[CalendarEntry]
    total: int
    day_counts: dict[str, int]
    truncated: bool


class AssigneeRead(BaseModel):
    id: int
    username: str

File: frontend/lib/features/tasks/domain/work_item.dart
enum WorkItemType { task, order, realization, reminder, event }

enum WorkItemStatus { todo, inProgress, completed, cancelled }

enum WorkItemPriority { low, normal, high, urgent }

class WorkAssignee {
  const WorkAssignee({required this.id, required this.username});
  final int id;
  final String username;
  factory WorkAssignee.fromJson(Map<String, dynamic> json) => WorkAssignee(
    id: json['id'] as int,
    username: json['username'].toString(),
  );
}

extension WorkItemTypePresentation on WorkItemType {
  String get api => name == 'inProgress' ? 'in_progress' : name;
  String get label => switch (this) {
    WorkItemType.task => 'Zadanie',
    WorkItemType.order => 'Zlecenie',
    WorkItemType.realization => 'Realizacja',
    WorkItemType.reminder => 'Przypomnienie',
    WorkItemType.event => 'Wydarzenie',
  };
}

class WorkItem {
  const WorkItem({
    required this.id,
    required this.type,
    required this.title,
    required this.status,
    required this.priority,
    required this.version,
    this.allDay = false,
    this.description,
    this.startAt,
    this.dueAt,
    this.assigneeUserId,
    this.assigneeDisplay,
    this.clientId,
    this.clientName,
    this.projectId,
    this.partyName,
    this.deletedAt,
    this.timezoneName,
  });
  final int id, version;
  final WorkItemType type;
  final String title;
  final String? description;
  final bool allDay;
  final String? timezoneName;
  final DateTime? startAt, dueAt, deletedAt;
  final WorkItemStatus status;
  final WorkItemPriority priority;
  final int? assigneeUserId, clientId, projectId;
  final String? assigneeDisplay, clientName, partyName;
  factory WorkItem.fromJson(Map<String, dynamic> j) => WorkItem(
    id: j['id'] as int,
    type: WorkItemType.values.firstWhere((v) => v.name == j['item_type']),
    title: j['title'].toString(),
    description: j['description']?.toString(),
    startAt: DateTime.tryParse(j['start_at']?.toString() ?? ''),
    dueAt: DateTime.tryParse(j['due_at']?.toString() ?? ''),
    allDay: j['all_day'] == true,
    timezoneName: j['timezone_name']?.toString(),
    status: WorkItemStatus.values.firstWhere(
      (v) => v.name.replaceAll('inProgress', 'in_progress') == j['status'],
    ),
    priority: WorkItemPriority.values.firstWhere(
      (v) => v.name == j['priority'],
    ),
    assigneeUserId: j['assignee_user_id'] as int?,
    assigneeDisplay: j['assignee_display']?.toString(),
    clientId: j['client_id'] as int?,
    clientName: j['client_name']?.toString(),
    projectId: j['project_id'] as int?,
    partyName: j['party_name']?.toString(),
    deletedAt: DateTime.tryParse(j['deleted_at']?.toString() ?? ''),
    version: j['version'] as int,
  );
}

class CalendarEntry {
  const CalendarEntry({
    required this.id,
    required this.kind,
    required this.type,
    required this.title,
    required this.start,
    required this.end,
    required this.status,
    this.priority,
    this.clientId,
  });
  final int id;
  final String kind, type, title, status;
  final DateTime start, end;
  final String? priority;
  final int? clientId;
  bool covers(DateTime day) {
    final d = DateTime(day.year, day.month, day.day);
    return !d.isBefore(DateTime(start.year, start.month, start.day)) &&
        !d.isAfter(DateTime(end.year, end.month, end.day));
  }

  factory CalendarEntry.fromJson(Map<String, dynamic> j) => CalendarEntry(
    id: j['entity_id'] as int,
    kind: j['entity_kind'].toString(),
    type: j['item_type'].toString(),
    title: j['title'].toString(),
    start: DateTime.parse(j['start'].toString()),
    end: DateTime.parse(j['end'].toString()),
    status: j['status'].toString(),
    priority: j['priority']?.toString(),
    clientId: j['client_id'] as int?,
  );
}

class CalendarMonthData {
  const CalendarMonthData({
    required this.year,
    required this.month,
    required this.items,
    required this.total,
    required this.dayCounts,
    required this.truncated,
  });
  final int year, month, total;
  final List<CalendarEntry> items;
  final Map<String, int> dayCounts;
  final bool truncated;
  factory CalendarMonthData.fromJson(Map<String, dynamic> j) =>
      CalendarMonthData(
        year: j['year'] as int,
        month: j['month'] as int,
        items: (j['items'] as List)
            .cast<Map<String, dynamic>>()
            .map(CalendarEntry.fromJson)
            .toList(),
        total: j['total'] as int,
        dayCounts: (j['day_counts'] as Map<String, dynamic>).map(
          (k, v) => MapEntry(k, v as int),
        ),
        truncated: j['truncated'] as bool,
      );
}

class AbsenceRequestItem {
  const AbsenceRequestItem({
    required this.id,
    required this.requesterUserId,
    required this.type,
    required this.start,
    required this.end,
    required this.status,
    required this.version,
    this.requesterDisplay,
    this.note,
  });
  final int id, version;
  final int requesterUserId;
  final String type, status;
  final DateTime start, end;
  final String? requesterDisplay;
  final String? note;
  factory AbsenceRequestItem.fromJson(Map<String, dynamic> j) =>
      AbsenceRequestItem(
        id: j['id'] as int,
        requesterUserId: j['requester_user_id'] as int,
        type: j['absence_type'].toString(),
        start: DateTime.parse(j['start_date'].toString()),
        end: DateTime.parse(j['end_date'].toString()),
        status: j['status'].toString(),
        version: j['version'] as int,
        requesterDisplay: j['requester_display']?.toString(),
        note: j['note']?.toString(),
      );
}

class WorkItemNote {
  const WorkItemNote({
    required this.id,
    required this.workItemId,
    required this.text,
    required this.version,
  });
  final int id;
  final int workItemId;
  final String text;
  final int version;
  factory WorkItemNote.fromJson(Map<String, dynamic> json) => WorkItemNote(
    id: json['id'] as int,
    workItemId: json['work_item_id'] as int,
    text: json['text'].toString(),
    version: json['version'] as int,
  );
}

class WorkItemDocument {
  const WorkItemDocument({
    required this.id,
    required this.workItemId,
    required this.documentId,
    required this.filename,
    required this.contentType,
    required this.fileSize,
    required this.sourceType,
    required this.createdAt,
    this.noteId,
    this.capturedAt,
  });

  final int id;
  final int workItemId;
  final int? noteId;
  final int documentId;
  final String filename;
  final String contentType;
  final int fileSize;
  final String sourceType;
  final DateTime? capturedAt;
  final DateTime createdAt;

  factory WorkItemDocument.fromJson(Map<String, dynamic> json) =>
      WorkItemDocument(
        id: json['id'] as int,
        workItemId: json['work_item_id'] as int,
        noteId: json['note_id'] as int?,
        documentId: json['document_id'] as int,
        filename: json['filename'].toString(),
        contentType: json['content_type'].toString(),
        fileSize: json['file_size'] as int,
        sourceType: json['source_type'].toString(),
        capturedAt: DateTime.tryParse(json['captured_at']?.toString() ?? ''),
        createdAt: DateTime.parse(json['created_at'].toString()),
      );
}
